File: pytecode/types/info.py
from dataclasses import dataclass
from typing import List

from . import attributes, constants, constant_pool


@dataclass
class FieldInfo:
    access_flags: constants.FieldAccessFlag
    name_index: int
    descriptor_index: int
    attributes_count: int
    attributes: List[attributes.AttributeInfo]


@dataclass
class MethodInfo:
    access_flags: constants.MethodAccessFlag
    name_index: int
    descriptor_index: int
    attributes_count: int
    attributes: List[attributes.AttributeInfo]


@dataclass
class ClassInfo:
    magic: int
    minor_version: int
    major_version: int
    contant_pool_count: int
    constant_pool: List[constant_pool.ConstantPoolInfo]
    access_flags: constants.ClassAccessFlag
    this_class: int
    super_class: int
    interfaces_count: int
    interfaces: List[int]
    fields_count: int
    fields: List[FieldInfo]
    methods_count: int
    methods: List[MethodInfo]
    attributes_count: int
    attributes: List[attributes.AttributeInfo]

File: crates/pytecode-engine/src/raw/info.rs
use crate::constants::{ClassAccessFlags, FieldAccessFlags, MethodAccessFlags};
use crate::raw::attributes::AttributeInfo;
use crate::raw::constant_pool::ConstantPoolEntry;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FieldInfo {
    pub access_flags: FieldAccessFlags,
    pub name_index: u16,
    pub descriptor_index: u16,
    pub attributes: Vec<AttributeInfo>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MethodInfo {
    pub access_flags: MethodAccessFlags,
    pub name_index: u16,
    pub descriptor_index: u16,
    pub attributes: Vec<AttributeInfo>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ClassFile {
    pub magic: u32,
    pub minor_version: u16,
    pub major_version: u16,
    pub constant_pool: Vec<Option<ConstantPoolEntry>>,
    pub access_flags: ClassAccessFlags,
    pub this_class: u16,
    pub super_class: u16,
    pub interfaces: Vec<u16>,
    pub fields: Vec<FieldInfo>,
    pub methods: Vec<MethodInfo>,
    pub attributes: Vec<AttributeInfo>,
}
